### DISCUSSION TITLE
バックエンドの多言語対応の仕様を考えよう

### DISCUSSION THREAD

**Discussion by kakkokari-gtyih** [misskey-dev/misskey#14306] [General]
メール送信やサーバーサイドエラーなどでバックエンドのi18nは重要だけど、かと言ってサーバーサイドでしか使わない文言をフロントエンドに流すのも無駄なので具体的にどういう実装（ディレクトリ構造）にするか考えたい

関連Issue:

- https://github.com/misskey-dev/misskey/issues/11060
- https://github.com/misskey-dev/misskey/issues/13472

**Comment by kakkokari-gtyih** [comment]
> メール送信やサーバーサイドエラーなどでバックエンドのi18nは重要だけど、かと言ってサーバーサイドでしか使わない文言をフロントエンドに流すのも無駄なので具体的にどういう実装（ディレクトリ構造）にするか考えたい

ご意見募集中

**Comment by samunohito** [comment]
そもそも、ニーズがあるのかしら…

**Reply by Sayamame-beans** [reply]
どちらかといえば、バックエンドのエラー表示がそのまま表に出てきていること(何箇所かあったはず?)が問題なのではないかと推測…
(クライアント側ではなくサーバー側のログにだけ出るものに関しては、問題無いのではと個人的には思ったりします。Misskeyとしての第一言語が日本語なのであればそちらに移るのはアリかも?)

そのため、バックエンドではi18nせずに、返ってきた内容を元に、特に必要なものに関しては(エラーメッセージの一種として、他のテキストと同じ様に)フロント側でi18nして表示してあげた方が良さそう…ですかね?

**Reply by kakkokari-gtyih** [reply]
- https://github.com/misskey-dev/misskey/issues/11060
- https://github.com/misskey-dev/misskey/issues/13472

**Comment by samunohito** [comment]
バックエンドのi18n化をする方法を考える前に、

- なぜバックエンドにi18nが必要なのか（課題は何処か）
- どこをi18n化すべきなのか

を整理したほうが良さそう…？

**Reply by kakkokari-gtyih** [reply]
- メール系の翻訳（メールを使用する機能の実装がいくつか提案されているので取り組んだほうが良さそう）
- エラーページ系の翻訳（サーバーエラー(fastifyのエラーハンドラ)の場合はフロントエンドでケアできない可能性がある）

**Comment by GrapeApple0** [comment]
個人的にはクライアントがまず見ないであろう部分を多言語化してもメリットはそんなに感じられませんし、エラーの部分はフロントエンドでカバーするかdocsにまとめればいいと思っています
メールについてはsignup時に言語を選択できるようにするか、クライアントの使用している言語をバックエンドに送信してそれに合わせてメールを送信すればいいと思います

**Reply by kakkokari-gtyih** [reply]
> メールについてはsignup時に言語を選択できるようにするか、クライアントの使用している言語をバックエンドに送信してそれに合わせてメールを送信すればいいと思います

メール用の文言をどういうふうに保管するかを決めあぐねているんですよね…
メールの文面をフロント向けの言語ファイルに含むのって無駄じゃないですか

**Reply by GrapeApple0** [reply]
バックエンド直下に`email-template`というディレクトリを作ってそこにhtmlなどのテンプレートを置いてそれを都度呼び出すやりかたがあるかなって思いました
